src/rocksdb.rs

```diff
@@ -2037,6 +2039,87 @@ impl Drop for BackupEngine {
     }
 }
 
+// SstFileReader is used to read sst files that are generated by DB or
+// SstFileWriter.
+pub struct SstFileReader {
+    inner: *mut crocksdb_ffi::SstFileReader,
+    _opt: ColumnFamilyOptions,
+}
+
+unsafe impl Send for SstFileReader {}
+
+impl SstFileReader {
+    /// Creates a new SstFileReader.
+    pub fn new(opt: ColumnFamilyOptions) -> Self {
+        unsafe {
+            Self {
+                inner: crocksdb_ffi::crocksdb_sstfilereader_create(opt.inner),
+                _opt: opt,
+            }
+        }
+    }
+
+    /// Opens a local SST file for reading.
+    pub fn open(&mut self, name: &str) -> Result<(), String> {
+        let path =
+            CString::new(name.to_owned()).map_err(|e| format!("invalid path {}: {:?}", name, e))?;
+        unsafe {
+            Ok(ffi_try!(crocksdb_sstfilereader_open(
+                self.inner,
+                path.as_ptr()
+            )))
+        }
+    }
+
+    pub fn iter(&self) -> DBIterator<&Self> {
+        self.iter_opt(ReadOptions::new())
+    }
+
+    pub fn iter_opt(&self, readopts: ReadOptions) -> DBIterator<&Self> {
+        unsafe {
+            DBIterator {
+                inner: crocksdb_ffi::crocksdb_sstfilereader_new_iterator(
+                    self.inner,
+                    readopts.get_inner(),
+                ),
+                _db: self,
+                _readopts: readopts,
+            }
+        }
+    }
+
+    pub fn read_table_properties<F: FnOnce(&TableProperties)>(&self, mut action: F) {
+        extern "C" fn callback<F: FnOnce(&TableProperties)>(
+            ctx: *mut c_void,
+            ptr: *const crocksdb_ffi::DBTableProperties,
+        ) {
+            unsafe {
+                let caller = ptr::read(ctx as *mut F);
+                caller(TableProperties::from_ptr(ptr));
+            }
+        }
+
+        unsafe {
+            crocksdb_ffi::crocksdb_sstfilereader_read_table_properties(
+                self.inner,
+                &mut action as *mut F as *mut c_void,
+                callback::<F>,
+            );
+            mem::forget(action);
```

Comment:
Why is this forget?

Reply:
@BusyJay the variable `action` as cast into a `*mut c_void` with ownership into `crocksdb_sstfilereader_read_table_properties` on line 2105 as `ctx`, which then jump to `c.cc:3534` invoking `callback::<F>()`, and finally the `ctx` is recovered as an `F` on line 2097 inside `callback::<F>()`, which will be freed there.

if we don't call `mem::forget`, the `action` will be freed twice, once after line 2098 and once here.

we have to pass with ownership because we want `F: FnOnce`, not `F: FnMut`.